Training hangs with sample_packing for specific datasets [multi-GPU]

### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Training should run normally.

### Current behaviour

With specific datasets and a multi-GPU setup (2*RTX 4090s in my case), training hangs (after the first training step or after the fist epoch) until a timeout error occurs.

The bug only happens:

- with 2 (or more?) GPUs
- with specific datasets (the average/maximum number of tokens per example seems to be important)
- with `sequence_len: 4096` (2048 runs fine)
- with `sample_packing: true`

With `micro_batch_size: 1` instead of 2 (which is also used for `sample_packing_seq_len_mulitpler`!), training starts fine but hangs after epoch 1 has finished.

When training hangs, both GPUs are stuck below 1/3 TWP (~90W in my case) and VRAM is still allocated.
P2P settings like `NCCL_P2P_DISABLE=1` don't change anything. 

### Steps to reproduce

Steps to reproduce (with config from [Llama-2/examples/lora.yml](https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/examples/llama-2/lora.yml) below.
Minimal example dataset to reproduce the issue is here [jphme/test_ds_2gpu_axolotl_bug](https://huggingface.co/datasets/jphme/test_ds_2gpu_axolotl_bug).

```bash
#works fine
accelerate launch scripts/finetune.py examples/llama-2/lora.yml

#hangs with 2gpus, nothing except datasegt changed
accelerate launch scripts/finetune.py examples/llama-2/lora_w_dataset_above.yml
```

Reducing sequence_len to 2048 or disabling sample_packing prevents the hanging; when setting `micro_batch_size: 1` training runs fine until epoch 1 has finished and then hangs. 


### Possible solution

I suspect that the bug has something todo with the `sample_packing` implementation and is only triggered for examples with many tokens:

Numer of tokens of the original Llama-2 lora.yml dataset `mhenrichsen/alpaca_2k_test` which runs fine:
```bash
count    2000.000000
mean      217.408000
std       146.209145
min        49.000000
25%        95.000000
50%       167.000000
75%       313.000000
max       762.000000
dtype: float64
```

Token length of my custom dataset:
```
count      25.000000
mean     1805.000000
std       603.574629
min       636.000000
25%      1274.000000
50%      1885.000000
75%      2184.000000
max      2991.000000
dtype: float64
```

As training hangs only for long sequence lengths, multiple GPUs and dataset with examples with many tokens, it should be possible to identify the root cause; unfortunately im out of my depth with the sample_packing code. 


### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.9.17 (via Dockerfile)

### axolotl branch-commit

main/f144e98a3259fafcb6d359147b756816333a227e

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

cc @winglian

How to do that? Never used pyspy ;-) .

But I got along further now, it seems to be some weird NCCL issue with distributed dataloaders.
With NCCL Async error handling and Accelerate debug i got the following output after the timeout:

```
ProcessGroupNCCL.cpp:828] [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=42, OpType=BROADCAST, Timeout(ms)=1800000) ran for 1809852 milliseconds before timing out.
[E ProcessGroupNCCL.cpp:455] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data.
[E ProcessGroupNCCL.cpp:460] To avoid data inconsistency, we are taking the entire process down.
terminate called after throwing an instance of 'std::runtime_error'
  what():  [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=42, OpType=BROADCAST, Timeout(ms)=1800000) ran for 1809852 milliseconds before timing out.
```

Going throgh almost all [NCCL settings](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-p2p-level), I found that the setting
`-e NCCL_BUFFSIZE=536870912` (=512 MB, default is 4 MB) fixes the problenm and training runs through, even with  `micro_batch_size: 2` .

Do you know some source discussing these settings and their tradeoffs (do I have 512MB less VRAM per GPU now?) ?- I found basically nothing with Google and it was pure trial&error.....

EDIT: Now training completes, but I get some strange AssertionError afterwards (will rebuild container to be sure that this has no other sources):

```bash
{'loss': 5.7284, 'learning_rate': 4.8943483704846475e-06, 'epoch': 7.6}
{'loss': 5.7308, 'learning_rate': 0.0, 'epoch': 8.0}
100%|███████████████████████████████████████████████████████████████████████| 20/20 [04:40<00:00, 14.45s/it]/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer_pt_utils.py:289: FutureWarning: SequentialDistributedSampler is deprecated and will be removed in v5 of Transformers.
  warnings.warn(
/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer_pt_utils.py:289: FutureWarning: SequentialDistributedSampler is deprecated and will be removed in v5 of Transformers.
  warnings.warn(
[2023-08-28 16:08:33,192] [INFO] [axolotl.utils.dataloader._len_est:262] [PID:98] packing_efficiency_estimate: 0.86 total_num_tokens per device: 979
[2023-08-28 16:08:33,192] [INFO] [axolotl.utils.dataloader._len_est:262] [PID:99] packing_efficiency_estimate: 0.86 total_num_tokens per device: 979
[2023-08-28 16:08:33,197] [INFO] [axolotl.utils.dataloader.generate_batches:181] [PID:98] generating packed batches
[2023-08-28 16:08:33,197] [INFO] [axolotl.utils.dataloader.generate_batches:181] [PID:99] generating packed batches
Traceback (most recent call last):
  File "/workspace/axolotl/scripts/finetune.py", line 326, in <module>
    fire.Fire(train)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
Traceback (most recent call last):
  File "/workspace/axolotl/scripts/finetune.py", line 326, in <module>
    component = fn(*varargs, **kwargs)
  File "/workspace/axolotl/scripts/finetune.py", line 303, in train
    fire.Fire(train)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    trainer.train(resume_from_checkpoint=resume_from_checkpoint)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1555, in train
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    return inner_training_loop(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1929, in _inner_training_loop
    component = fn(*varargs, **kwargs)
  File "/workspace/axolotl/scripts/finetune.py", line 303, in train
    trainer.train(resume_from_checkpoint=resume_from_checkpoint)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1555, in train
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2256, in _maybe_log_save_evaluate
    return inner_training_loop(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1929, in _inner_training_loop
    metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2972, in evaluate
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2256, in _maybe_log_save_evaluate
    metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2972, in evaluate
    output = eval_loop(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 3151, in evaluation_loop
        output = eval_loop(for step, inputs in enumerate(dataloader):

  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 3151, in evaluation_loop
  File "/workspace/axolotl/src/axolotl/utils/dataloader.py", line 214, in __iter__
    all_batches, _ = self.generate_batches(set_stats=True)
  File "/workspace/axolotl/src/axolotl/utils/dataloader.py", line 183, in generate_batches
    indices = [idx for idx in self.sampler]
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer_pt_utils.py", line 318, in __iter__
    assert (
AssertionError: Indices length 2 and total size 4 mismatched
    for step, inputs in enumerate(dataloader):
  File "/workspace/axolotl/src/axolotl/utils/dataloader.py", line 214, in __iter__
    all_batches, _ = self.generate_batches(set_stats=True)
  File "/workspace/axolotl/src/axolotl/utils/dataloader.py", line 183, in generate_batches
    indices = [idx for idx in self.sampler]
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer_pt_utils.py", line 318, in __iter__
    assert (
AssertionError: Indices length 2 and total size 4 mismatched
```

--- a/examples/code-llama/13b/lora.yml
+++ b/examples/code-llama/13b/lora.yml
@@ -17,6 +17,7 @@ output_dir: ./lora-out
 
 sequence_len: 100000
 sample_packing: true
+pad_to_sequence_len: true
 
 adapter: lora
 lora_model_dir:

--- a/examples/code-llama/13b/qlora.yml
+++ b/examples/code-llama/13b/qlora.yml
@@ -20,6 +20,7 @@ lora_model_dir:
 
 sequence_len: 100000
 sample_packing: true
+pad_to_sequence_len: true
 
 lora_r: 32
 lora_alpha: 16

--- a/examples/code-llama/34b/lora.yml
+++ b/examples/code-llama/34b/lora.yml
@@ -17,6 +17,7 @@ output_dir: ./lora-out
 
 sequence_len: 100000
 sample_packing: true
+pad_to_sequence_len: true
 
 adapter: lora
 lora_model_dir:

--- a/examples/code-llama/34b/qlora.yml
+++ b/examples/code-llama/34b/qlora.yml
@@ -20,6 +20,7 @@ lora_model_dir:
 
 sequence_len: 100000
 sample_packing: true
+pad_to_sequence_len: true
 
 lora_r: 32
 lora_alpha: 16

--- a/examples/code-llama/7b/lora.yml
+++ b/examples/code-llama/7b/lora.yml
@@ -17,6 +17,7 @@ output_dir: ./lora-out
 
 sequence_len: 100000
 sample_packing: true
+pad_to_sequence_len: true
 
 adapter: lora
 lora_model_dir:

--- a/examples/code-llama/7b/qlora.yml
+++ b/examples/code-llama/7b/qlora.yml
@@ -20,6 +20,7 @@ lora_model_dir:
 
 sequence_len: 100000
 sample_packing: true
+pad_to_sequence_len: true
 
 lora_r: 32
 lora_alpha: 16

--- a/examples/llama-2/lora.yml
+++ b/examples/llama-2/lora.yml
@@ -17,6 +17,7 @@ output_dir: ./lora-out
 
 sequence_len: 4096
 sample_packing: true
+pad_to_sequence_len: true
 
 adapter: lora
 lora_model_dir:

--- a/examples/llama-2/qlora.yml
+++ b/examples/llama-2/qlora.yml
@@ -20,6 +20,7 @@ lora_model_dir:
 
 sequence_len: 4096
 sample_packing: true
+pad_to_sequence_len: true
 
 lora_r: 32
 lora_alpha: 16

--- a/examples/llama-2/relora.yml
+++ b/examples/llama-2/relora.yml
@@ -20,6 +20,7 @@ lora_model_dir:
 
 sequence_len: 4096
 sample_packing: true
+pad_to_sequence_len: true
 
 lora_r: 8
 lora_alpha: 16

--- a/src/axolotl/utils/config.py
+++ b/src/axolotl/utils/config.py
@@ -97,6 +97,11 @@ def validate_config(cfg):
             )
         )
 
+    if cfg.sample_packing and not cfg.pad_to_sequence_len:
+        LOG.warning(
+            "`pad_to_sequence_len: true` is recommended when using sample_packing"
+        )
+
     if cfg.gradient_accumulation_steps and cfg.batch_size:
         raise ValueError(
             "please set only one of gradient_accumulation_steps or batch_size"

--- a/tests/test_validation.py
+++ b/tests/test_validation.py
@@ -330,6 +330,20 @@ class ValidationTest(unittest.TestCase):
 
         cfg = DictDefault(
             {
+                "sample_packing": True,
+                "pad_to_sequence_len": None,
+            }
+        )
+        with self._caplog.at_level(logging.WARNING):
+            validate_config(cfg)
+            assert any(
+                "`pad_to_sequence_len: true` is recommended when using sample_packing"
+                in record.message
+                for record in self._caplog.records
+            )
+
+        cfg = DictDefault(
+            {
                 "max_packed_sequence_len": 2048,
                 "sample_packing": True,
             }

